src/arraymancer/tensor/selectors.nim

```diff
@@ -179,6 +178,82 @@ proc masked_fill*[T](t: var Tensor[T], mask: openArray, value: T) =
     return
   t.masked_fill(mask.toTensor(), value)
 
+proc masked_fill*[T](t: var Tensor[T], mask: Tensor[bool], value: Tensor[T]) =
+  ## For the index of each element of t.
+  ## Fill the elements at ``t[index]`` with the ``value``
+  ## if their corresponding ``mask[index]`` is true.
+  ## If not they are untouched.
+  ##
+  ## Example:
+  ##
+  ##   t.masked_fill(t > 0, -1)
+  ##
+  ## or alternatively:
+  ##
+  ##   t.masked_fill(t > 0): -1
+  if t.size == 0 or mask.size == 0:
+    return
+  check_elementwise(t, mask)
+
+  # Due to requiring unnecessary assigning `x` for a `false` mask
+  # apply2_inline is a bit slower for very sparse mask.
+  # As this is a critical operation, especially on dataframes, we use the lower level construct.
+  #
+  # t.apply2_inline(mask):
+  #   if y:
+  #     value
+  #   else:
+  #     x
+
+  # We need to protect against the case in which we run out of values to
+  # fill the tensor with, which we cannot tell in advance without counting
+  # the number of true values in the mask (which would be expensive)
+  var lock: Lock
+  initLock(lock)
+  var too_few_values = false
+
+  omp_parallel_blocks(block_offset, block_size, t.size):
+    var n = block_offset
+    for tElem, maskElem in mzip(t, mask, block_offset, block_size):
+      if maskElem:
+        if n >= value.size:
+          withLock(lock):
+            # The lock protection is technically unnecessary but it is good form
+            too_few_values = true
+          break
+        tElem = value[n]
+        inc n
+  if too_few_values:
+    let error_msg = "masked_fill error: the size of the value tensor (" & $value.size &
+      ") is smaller than the number of true elements in the mask"
+    when not(compileOption("mm", "arc") or compileOption("mm", "orc")):
```

Comment:
An alterative that could work for older Nim would be:

```suggestion
    when not defined(gcDestructors)
```

but I don't mind to just drop 1.4.

Reply:
I can try this change. In any case I think dropping 1.4 makes sense.